In [2]:
import pandas as pd
import numpy as np

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

In [3]:
df = pd.read_csv("D:\Projects\Funnel-Analysis\data\messy_data.csv")

<>:1: SyntaxWarning: "\P" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\P"? A raw string is also an option.
<>:1: SyntaxWarning: "\P" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\P"? A raw string is also an option.
C:\Users\slaks\AppData\Local\Temp\ipykernel_3788\3321552346.py:1: SyntaxWarning: "\P" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\P"? A raw string is also an option.
  df = pd.read_csv("D:\Projects\Funnel-Analysis\data\messy_data.csv")


In [4]:
df.head()

,user_id,session_id,event_name,event_timestamp,product_id,category,price,device_type,traffic_source,country
0,1333.0,sess_ebdc6c7379e3,addtocart,2024-03-10 22:22:00,PROD_7753,Beauty,16466.98,tablet,organic,Australia
1,3249.0,sess_226b171c312b,AddToCart,2024-03-05 06:10:00,PROD_8143,Beauty,$73.4,mobile,organic,Brazil
2,3284.0,sess_129ae99a4cbe,Add_To_Cart,2024-03-23 21:30:00,PROD_6504,Books,78.27,mobile,organic,Canada
3,2435.0,sess_0b41ee99a20d,ADD_TO_CART,2024/06/24 06:17,PROD_5364,Automotive,735.98,tablet,ads,USA
4,1241.0,sess_9e731e3492f7,AddToCart,18/01/2024 01:36,PROD_9459,Clothing,280.45,mobile,organic,Germany


In [5]:
df.shape

(24312, 10)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 24312 entries, 0 to 24311
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   user_id          24134 non-null  float64
 1   session_id       24114 non-null  str    
 2   event_name       24312 non-null  str    
 3   event_timestamp  24312 non-null  str    
 4   product_id       21866 non-null  str    
 5   category         21650 non-null  str    
 6   price            20367 non-null  str    
 7   device_type      22513 non-null  str    
 8   traffic_source   22276 non-null  str    
 9   country          22498 non-null  str    
dtypes: float64(1), str(9)
memory usage: 1.9 MB


In [8]:
df.describe(include='all')

,user_id,session_id,event_name,event_timestamp,product_id,category,price,device_type,traffic_source,country
count,24134.000000,24114,24312,24312,21866,21650,20367,22513,22276,22498
unique,NaN,8397,40,23617,6821,20,10903,3,11,12
top,NaN,sess_d1dbfff9a029,view_item,02/02/2024 21:50,PROD_6304,Beauty,33.86,desktop,referral,USA
freq,NaN,8,1285,3,17,2161,11,7590,5311,1918
mean,3015.714842,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,1157.296998,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,1001.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,1999.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,3034.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,4025.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [9]:
df.isnull().sum()

user_id             178
session_id          198
event_name            0
event_timestamp       0
product_id         2446
category           2662
price              3945
device_type        1799
traffic_source     2036
country            1814
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(355)

In [11]:
df['event_name'].unique()

<StringArray>
[     'addtocart',      'AddToCart',    'Add_To_Cart',    'ADD_TO_CART',
     'null_event',           'View',          'click',      'CHECK_OUT',
           'veiw',       'cart_add',           'view',      'check_out',
           'VIEW',          'login',    'error_event',       'checkout',
    'add_to_cart',      'view_item',    'add-to-cart',   'wishlist_add',
         'search',      'page_view',      'page_exit',    'promo_click',
          'hover',       'Purchase',         'viewed',       'CHECKOUT',
         'scroll',        'chekout',         'logout',        'UNKNOWN',
            'buy',   'order_placed', 'begin_checkout',        'purchse',
       'Checkout',       'purchase',       'PURCHASE',      'purchased']
Length: 40, dtype: str

In [12]:
df['user_id'] = df['user_id'].astype('Int64')

In [13]:
df['event_timestamp'] = pd.to_datetime(df['event_timestamp'], errors='coerce')

In [14]:
df['event_timestamp'].isnull().sum()

np.int64(9655)

In [15]:
df['price'] = pd.to_numeric(df['price'], errors='coerce')

In [16]:
df['price'].isnull().sum()
df['price'].describe()

count    19869.000000
mean      1384.483319
std       7721.137151
min          0.030000
25%         60.610000
50%        187.510000
75%        555.130000
max      99618.320000
Name: price, dtype: float64

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 24312 entries, 0 to 24311
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   user_id          24134 non-null  Int64         
 1   session_id       24114 non-null  str           
 2   event_name       24312 non-null  str           
 3   event_timestamp  14657 non-null  datetime64[us]
 4   product_id       21866 non-null  str           
 5   category         21650 non-null  str           
 6   price            19869 non-null  float64       
 7   device_type      22513 non-null  str           
 8   traffic_source   22276 non-null  str           
 9   country          22498 non-null  str           
dtypes: Int64(1), datetime64[us](1), float64(1), str(7)
memory usage: 1.9 MB


In [18]:
df.isnull().sum().sort_values(ascending=False)

event_timestamp    9655
price              4443
category           2662
product_id         2446
traffic_source     2036
country            1814
device_type        1799
session_id          198
user_id             178
event_name            0
dtype: int64

In [22]:
df = df.dropna(subset=['user_id', 'session_id', 'event_name', 'event_timestamp'])

In [23]:
df.shape

(14423, 10)

In [24]:
df.isnull().sum()

user_id               0
session_id            0
event_name            0
event_timestamp       0
product_id         1479
category           1616
price              2694
device_type        1063
traffic_source     1222
country            1082
dtype: int64

In [25]:
df['event_name'].value_counts()

event_name
page_view         769
viewed            746
VIEW              740
view              721
view_item         715
veiw              673
View              658
cart_add          471
add_to_cart       460
Add_To_Cart       447
add-to-cart       433
AddToCart         431
ADD_TO_CART       427
addtocart         420
promo_click       344
wishlist_add      342
login             333
click             332
scroll            331
page_exit         319
UNKNOWN           319
null_event        315
error_event       309
search            308
logout            307
hover             302
Checkout          267
checkout          252
check_out         247
CHECKOUT          242
begin_checkout    235
chekout           234
CHECK_OUT         233
purchase          126
buy               115
Purchase          105
order_placed      105
purchased         104
PURCHASE           97
purchse            89
Name: count, dtype: int64

In [26]:
df['event_name'] = df['event_name'].str.lower().str.strip()

In [27]:
df['event_name'].value_counts()

event_name
view              2119
add_to_cart       1334
addtocart          851
page_view          769
checkout           761
viewed             746
view_item          715
veiw               673
check_out          480
cart_add           471
add-to-cart        433
promo_click        344
wishlist_add       342
login              333
click              332
scroll             331
purchase           328
page_exit          319
unknown            319
null_event         315
error_event        309
search             308
logout             307
hover              302
begin_checkout     235
chekout            234
buy                115
order_placed       105
purchased          104
purchse             89
Name: count, dtype: int64

In [29]:
event_mapping = {
    'view': 'view',
    'veiw': 'view',

    'addtocart': 'add_to_cart',
    'add_to_cart': 'add_to_cart',
    'add to cart': 'add_to_cart',

    'purchase': 'purchase',
    'purchse': 'purchase'
}

In [30]:
df['event_name'] = df['event_name'].map(event_mapping)

In [31]:
df['event_name'].isnull().sum()

np.int64(9029)

In [32]:
df = df.dropna(subset=['event_name'])

In [33]:
df['event_name'].value_counts()

event_name
view           2792
add_to_cart    2185
purchase        417
Name: count, dtype: int64

In [34]:
df['event_name'].value_counts().head(20)

event_name
view           2792
add_to_cart    2185
purchase        417
Name: count, dtype: int64

In [35]:
cols = ['device_type', 'traffic_source', 'country']

for col in cols:
    df[col] = df[col].str.lower().str.strip()

In [38]:
df['device_type'].value_counts()


device_type
desktop    1779
mobile     1744
tablet     1725
Name: count, dtype: int64

In [39]:
df['traffic_source'].value_counts()

traffic_source
direct      1288
referral    1242
ads         1231
organic     1165
orgnaic       73
organik       71
refferal      58
add           53
Name: count, dtype: int64

In [40]:
df['country'].value_counts()

country
mexico       468
uae          456
canada       452
uk           452
germany      449
usa          441
india        440
france       428
japan        424
brazil       421
singapore    420
australia    398
Name: count, dtype: int64

In [41]:
df['traffic_source'] = df['traffic_source'].replace({
    'orgnaic': 'organic',
    'organik': 'organic',
    'refferal': 'referral',
    'add': 'ads'
})

In [42]:
df['traffic_source'].value_counts()

traffic_source
organic     1309
referral    1300
direct      1288
ads         1284
Name: count, dtype: int64

In [43]:
df.duplicated().sum()

np.int64(74)

In [44]:
df = df.drop_duplicates()

In [45]:
df.duplicated().sum()

np.int64(0)

In [46]:
df.head

<bound method NDFrame.head of        user_id         session_id   event_name     event_timestamp product_id  \
0         1333  sess_ebdc6c7379e3  add_to_cart 2024-03-10 22:22:00  PROD_7753   
1         3249  sess_226b171c312b  add_to_cart 2024-03-05 06:10:00  PROD_8143   
2         3284  sess_129ae99a4cbe  add_to_cart 2024-03-23 21:30:00  PROD_6504   
11        3201  sess_a8a5628c88ac         view 2024-08-19 03:28:00  PROD_3911   
13        3845  sess_3a42ccd01a44         view 2024-10-16 22:18:00  PROD_5277   
...        ...                ...          ...                 ...        ...   
24291     4992  sess_b7cf9859eeab  add_to_cart 2024-05-26 02:36:00  PROD_3375   
24295     4994  sess_7dc71fb396a8         view 2024-01-17 13:12:00  PROD_5970   
24299     4995  sess_ab255b78339c  add_to_cart 2024-04-02 18:40:00  PROD_9581   
24304     4997  sess_c500d9c64fbe  add_to_cart 2024-09-06 07:37:00  PROD_4054   
24306     4998  sess_e077283510e8         view 2024-09-15 07:21:00  PROD_8923  

In [47]:
df.head(10)

,user_id,session_id,event_name,event_timestamp,product_id,category,price,device_type,traffic_source,country
0,1333,sess_ebdc6c7379e3,add_to_cart,2024-03-10 22:22:00,PROD_7753,Beauty,16466.98,tablet,organic,australia
1,3249,sess_226b171c312b,add_to_cart,2024-03-05 06:10:00,PROD_8143,Beauty,NaN,mobile,organic,brazil
2,3284,sess_129ae99a4cbe,add_to_cart,2024-03-23 21:30:00,PROD_6504,Books,78.27,mobile,organic,canada
11,3201,sess_a8a5628c88ac,view,2024-08-19 03:28:00,PROD_3911,Home & Kitchn,NaN,mobile,ads,mexico
13,3845,sess_3a42ccd01a44,view,2024-10-16 22:18:00,PROD_5277,Toys,22905.96,tablet,direct,uk
14,1006,sess_15f92a0feaf4,view,2024-08-31 08:13:00,PROD_2269,Sports,288.28,tablet,direct,germany
16,1426,sess_18ddaddcd50b,view,2024-11-05 16:03:00,PROD_4323,Home & Kitchen,469.80,desktop,ads,india
20,3902,sess_90234722f7af,add_to_cart,2024-03-12 06:49:00,PROD_6997,Books,94675.81,mobile,direct,canada
21,1298,sess_ad1fe1558ca8,view,2024-07-27 18:07:00,PROD_8126,Books,18.65,mobile,organic,mexico
23,1213,sess_3caa92c3f87e,add_to_cart,2024-02-03 13:36:00,PROD_5914,Jewelry,1355.05,mobile,ads,uk


In [48]:
df.groupby('event_name')['price'].apply(lambda x: x.isnull().mean())

event_name
add_to_cart    0.075996
purchase       0.050971
view           0.074182
Name: price, dtype: float64

In [49]:
df = df[~((df['event_name'] == 'purchase') & (df['price'].isnull()))]

In [50]:
df.head(10)

,user_id,session_id,event_name,event_timestamp,product_id,category,price,device_type,traffic_source,country
0,1333,sess_ebdc6c7379e3,add_to_cart,2024-03-10 22:22:00,PROD_7753,Beauty,16466.98,tablet,organic,australia
1,3249,sess_226b171c312b,add_to_cart,2024-03-05 06:10:00,PROD_8143,Beauty,NaN,mobile,organic,brazil
2,3284,sess_129ae99a4cbe,add_to_cart,2024-03-23 21:30:00,PROD_6504,Books,78.27,mobile,organic,canada
11,3201,sess_a8a5628c88ac,view,2024-08-19 03:28:00,PROD_3911,Home & Kitchn,NaN,mobile,ads,mexico
13,3845,sess_3a42ccd01a44,view,2024-10-16 22:18:00,PROD_5277,Toys,22905.96,tablet,direct,uk
14,1006,sess_15f92a0feaf4,view,2024-08-31 08:13:00,PROD_2269,Sports,288.28,tablet,direct,germany
16,1426,sess_18ddaddcd50b,view,2024-11-05 16:03:00,PROD_4323,Home & Kitchen,469.80,desktop,ads,india
20,3902,sess_90234722f7af,add_to_cart,2024-03-12 06:49:00,PROD_6997,Books,94675.81,mobile,direct,canada
21,1298,sess_ad1fe1558ca8,view,2024-07-27 18:07:00,PROD_8126,Books,18.65,mobile,organic,mexico
23,1213,sess_3caa92c3f87e,add_to_cart,2024-02-03 13:36:00,PROD_5914,Jewelry,1355.05,mobile,ads,uk


In [51]:
df[df['event_name'] == 'purchase']['price'].isnull().sum()

np.int64(0)

In [52]:
df[df['price'].isnull()]['event_name'].value_counts()

event_name
view           204
add_to_cart    164
Name: count, dtype: int64

In [53]:
df[df['event_name'] == 'purchase']['price'].isnull().sum()

np.int64(0)

In [54]:
funnel = df.groupby('event_name')['user_id'].nunique()
funnel

event_name
add_to_cart    1748
purchase        380
view           2098
Name: user_id, dtype: int64

In [55]:
#Calculate Conversion Rates

views = funnel.get('view', 0)
add_to_cart = funnel.get('add_to_cart', 0)
purchase = funnel.get('purchase', 0)

view_to_cart = add_to_cart / views
cart_to_purchase = purchase / add_to_cart
overall_conversion = purchase / views

view_to_cart, cart_to_purchase, overall_conversion

(np.float64(0.8331744518589133),
 np.float64(0.21739130434782608),
 np.float64(0.1811248808388942))

In [57]:
#Drop-off Analysis

drop_view_to_cart = 1 - view_to_cart
drop_cart_to_purchase = 1 - cart_to_purchase

drop_view_to_cart, drop_cart_to_purchase

(np.float64(0.16682554814108674), np.float64(0.782608695652174))

In [58]:
#Flags

df['view_flag'] = (df['event_name'] == 'view').astype(int)
df['cart_flag'] = (df['event_name'] == 'add_to_cart').astype(int)
df['purchase_flag'] = (df['event_name'] == 'purchase').astype(int)

In [59]:
session_funnel = df.groupby('session_id').agg({
    'view_flag': 'max',
    'cart_flag': 'max',
    'purchase_flag': 'max'
}).reset_index()

In [60]:
session_funnel.head()

,session_id,view_flag,cart_flag,purchase_flag
0,sess_000907ab9f7b,1,0,0
1,sess_000b7cd9fd09,0,1,0
2,sess_001c5a984949,1,0,0
3,sess_002fb04a5b62,0,1,0
4,sess_0060341d76b8,1,0,0


In [61]:
total_sessions = session_funnel.shape[0]

view_sessions = session_funnel['view_flag'].sum()
cart_sessions = session_funnel['cart_flag'].sum()
purchase_sessions = session_funnel['purchase_flag'].sum()

view_sessions, cart_sessions, purchase_sessions

(np.int64(2750), np.int64(2158), np.int64(391))

In [62]:
view_to_cart = cart_sessions / view_sessions
cart_to_purchase = purchase_sessions / cart_sessions
overall_conversion = purchase_sessions / view_sessions

view_to_cart, cart_to_purchase, overall_conversion

(np.float64(0.7847272727272727),
 np.float64(0.18118628359592215),
 np.float64(0.1421818181818182))

In [64]:
session_funnel = session_funnel[
    (session_funnel['view_flag'] == 1) &
    (session_funnel['cart_flag'] <= session_funnel['view_flag']) &
    (session_funnel['purchase_flag'] <= session_funnel['cart_flag'])
]

In [65]:
session_device = df.groupby('session_id')['device_type'].first().reset_index()

In [66]:
session_funnel = session_funnel.merge(session_device, on='session_id', how='left')

In [67]:
session_funnel.head()

,session_id,view_flag,cart_flag,purchase_flag,device_type
0,sess_000907ab9f7b,1,0,0,mobile
1,sess_001c5a984949,1,0,0,desktop
2,sess_0060341d76b8,1,0,0,desktop
3,sess_006fd196c391,1,0,0,mobile
4,sess_008ca9f59a75,1,0,0,tablet


In [68]:
device_funnel = session_funnel.groupby('device_type').agg({
    'view_flag': 'sum',
    'cart_flag': 'sum',
    'purchase_flag': 'sum'
}).reset_index()

In [69]:
#Conversion rates
device_funnel['view_to_cart'] = device_funnel['cart_flag'] / device_funnel['view_flag']
device_funnel['cart_to_purchase'] = device_funnel['purchase_flag'] / device_funnel['cart_flag']
device_funnel['overall_conversion'] = device_funnel['purchase_flag'] / device_funnel['view_flag']

device_funnel

,device_type,view_flag,cart_flag,purchase_flag,view_to_cart,cart_to_purchase,overall_conversion
0,desktop,875,237,14,0.270857,0.059072,0.016000
1,mobile,907,235,16,0.259096,0.068085,0.017641
2,tablet,845,240,14,0.284024,0.058333,0.016568


In [70]:
session_traffic = df.groupby('session_id')['traffic_source'].first().reset_index()

session_funnel = session_funnel.merge(session_traffic, on='session_id', how='left')

In [71]:
#Aggregation
traffic_funnel = session_funnel.groupby('traffic_source').agg({
    'view_flag': 'sum',
    'cart_flag': 'sum',
    'purchase_flag': 'sum'
}).reset_index()

In [72]:
#Conversion rates
traffic_funnel['view_to_cart'] = traffic_funnel['cart_flag'] / traffic_funnel['view_flag']
traffic_funnel['cart_to_purchase'] = traffic_funnel['purchase_flag'] / traffic_funnel['cart_flag']
traffic_funnel['overall_conversion'] = traffic_funnel['purchase_flag'] / traffic_funnel['view_flag']

traffic_funnel

,traffic_source,view_flag,cart_flag,purchase_flag,view_to_cart,cart_to_purchase,overall_conversion
0,ads,644,185,16,0.287267,0.086486,0.024845
1,direct,644,172,10,0.267081,0.058140,0.015528
2,organic,645,180,13,0.279070,0.072222,0.020155
3,referral,664,174,5,0.262048,0.028736,0.007530


In [73]:
session_country = df.groupby('session_id')['country'].first().reset_index()

session_funnel = session_funnel.merge(session_country, on='session_id', how='left')

In [74]:
country_funnel = session_funnel.groupby('country').agg({
    'view_flag': 'sum',
    'cart_flag': 'sum',
    'purchase_flag': 'sum'
}).reset_index()

In [75]:
country_funnel['view_to_cart'] = country_funnel['cart_flag'] / country_funnel['view_flag']
country_funnel['cart_to_purchase'] = country_funnel['purchase_flag'] / country_funnel['cart_flag']
country_funnel['overall_conversion'] = country_funnel['purchase_flag'] / country_funnel['view_flag']

country_funnel

,country,view_flag,cart_flag,purchase_flag,view_to_cart,cart_to_purchase,overall_conversion
0,australia,209,42,0,0.200957,0.000000,0.000000
1,brazil,204,62,1,0.303922,0.016129,0.004902
2,canada,212,60,2,0.283019,0.033333,0.009434
3,france,215,54,4,0.251163,0.074074,0.018605
4,germany,234,79,5,0.337607,0.063291,0.021368
5,india,224,56,1,0.250000,0.017857,0.004464
6,japan,214,53,4,0.247664,0.075472,0.018692
7,mexico,245,65,4,0.265306,0.061538,0.016327
8,singapore,208,63,3,0.302885,0.047619,0.014423
9,uae,236,66,8,0.279661,0.121212,0.033898


### 📱 Device-Based Funnel Insights

* **Tablet users have the highest view-to-cart conversion (28.4%)**, indicating strong product engagement at the browsing stage.
* **Mobile users have the highest overall conversion (1.76%)**, slightly outperforming desktop and tablet, suggesting relatively better end-to-end performance.
* **Desktop shows the lowest cart-to-purchase conversion (5.9%)**, indicating possible friction during checkout despite decent initial engagement.
* Overall, conversion rates across all devices are quite low (~1.6%–1.7%), suggesting systemic issues in the funnel rather than device-specific problems.


### 🌐 Traffic Source Insights

* **Ads traffic has the highest overall conversion rate (2.48%)**, making it the most effective channel in terms of driving purchases.
* **Organic traffic performs consistently well (2.01%)**, indicating high user intent and quality traffic.
* **Referral traffic performs the worst (0.75%)**, with extremely low cart-to-purchase conversion (2.87%), suggesting poor traffic quality or mismatched audience targeting.
* **Direct traffic shows moderate performance (1.55%)**, possibly representing returning users but with room for improvement in conversion.


### 🌍 Country-Based Funnel Insights

* **USA (3.36%) and UAE (3.39%) have the highest overall conversion rates**, significantly outperforming other countries, indicating strong purchasing behavior.
* **Germany shows strong top-of-funnel engagement (33.7% view-to-cart)**, but relatively lower purchase conversion, suggesting drop-offs in later stages.
* **Australia has zero purchases despite user activity**, indicating a critical issue such as payment barriers, pricing, or lack of trust.
* **India and Brazil show high engagement but extremely low purchase conversion**, suggesting potential issues like pricing sensitivity or payment limitations.
* Countries like **Japan, UK, and France show balanced funnel performance**, indicating stable and consistent user behavior.


### 🔥 Key Business Takeaways

* The biggest drop-off occurs in the **final stage (cart to purchase)** across most segments, indicating checkout or payment-related friction.
* High-performing regions (USA, UAE) suggest where scaling efforts (ads, marketing) could yield better ROI.
* Low-performing segments (referral traffic, certain countries) highlight areas where targeting, UX, or localization strategies need improvement.


In [76]:
df.to_csv("../data/cleaned_data.csv", index=False)